src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java

```diff
@@ -91,8 +96,77 @@ public void failed(Throwable exc, Void v) {
         }
     }
 
-    /** Returns a file object for the given cache key. */
-    File getFileForKey(String key) {
-        return new File(mRootDirectorySupplier.get(), DiskBasedCacheUtility.getFilenameForKey(key));
+    /** Puts the cache entry with a specified key into the cache. */
+    @Override
+    public void put(final String key, Cache.Entry entry, final OnPutCompleteCallback callback) {
+        // If adding this entry would trigger a prune, but pruning would cause the new entry to be
+        // deleted, then skip writing the entry in the first place, as this is just churn.
+        // Note that we don't include the cache header overhead in this calculation for simplicity,
+        // so putting entries which are just below the threshold may still cause this churn.
+        if (mTotalSize + entry.data.length > mMaxCacheSizeInBytes
+                && entry.data.length
+                        > mMaxCacheSizeInBytes * DiskBasedCacheUtility.HYSTERESIS_FACTOR) {
```

Comment:
It's a good idea to check for this. However, we now have duplicate code in `pruneIfNeeded` and here, in the sense we're checking the same conditions in 2 different places. This is error prone: in the future, if one condition is updated and someone forgets to update the other, it would result in the 2 checks being out of sync and we could end up with not adding entries to the cache when it would have been totally fine.

Instead, I suggest to have your conditions defined in a shared place, most likely in `DiskBasedCacheUtility`. It's OK to split them into 2 methods (one for checking the max cache size, and one with the hysteresis factor), but at least you'd have them defined in one place which will prevent them from being out of sync. There would still be a risk that `pruneIfNeeded` is updated and its logic doesn't correspond to the check here, so it's not perfect, but the chances of getting out of sync are decreased.

Reply:
Yeah, I agree with that. I just made those changes.